src/components/recs/rec-upload-form.tsx

```diff
@@ -20,53 +19,93 @@ export type RecUploadFormProps = {
   setRecs: Dispatch<SetStateAction<any[]>>;
   filters: Filters;
 };
-export default function RecUploadForm({ setRecs, filters }: RecUploadFormProps) {
-  const [recFile, setRecFile] = useState(null);
+
+export default function RecUploadForm({
+  setRecs,
+  filters,
+}: RecUploadFormProps) {
+  const [recFile, setRecFile] = useState<File | null>(null);
   const [fileName, setFileName] = useState("");
+  const [fileType, setFileType] = useState("");
   const [isUploading, setIsUploading] = useState(false);
-  const { data: session } = useSession(); // get the client session
+  const { data: session } = useSession();
 
-  const handleFileChange = (e: any) => {
-    const selectedFile = e.target.files[0];
-    if (!selectedFile?.name) return;
-    setRecFile(selectedFile);
+  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const selectedFile = e.target.files?.[0];
+    if (selectedFile) {
+      setRecFile(selectedFile);
+      // Check if the file has a .mythrec extension
+      if (selectedFile.name.toLowerCase().endsWith(".mythrec")) {
+        setFileType("application/x-mythrec");
+      } else {
+        setFileType(selectedFile.type || "application/octet-stream");
+      }
+    }
   };
 
-  const handleFileNameChange = (e: any) => {
+  const handleFileNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
     setFileName(e.target.value);
   };
 
-  async function handleUploadFile(e: any): Promise<void> {
+  async function handleUploadFile(
+    e: React.FormEvent<HTMLFormElement>,
+  ): Promise<void> {
     e.preventDefault();
     if (!recFile) return;
     setIsUploading(true);
 
     try {
-      const formData = new FormData();
-      formData.append("file", recFile);
-      formData.append("gameTitle", fileName);
+      // Step 1: Get the pre-signed URL
+      const presignedUrlResponse = await fetch(
+        `/api/getPresignedUrl?fileName=${encodeURIComponent(recFile.name)}&fileType=${encodeURIComponent(fileType)}`,
+      );
+      if (!presignedUrlResponse.ok) {
+        throw new Error("Failed to get pre-signed URL");
+      }
+      const { signedUrl, key } = await presignedUrlResponse.json();
 
-      const response = await fetch("/api/recordedGames", {
+      // Step 2: Upload the file to S3
+      const uploadResponse = await fetch(signedUrl, {
+        method: "PUT",
+        body: recFile,
+        headers: {
+          "Content-Type": fileType,
+        },
+        mode: "cors",
+      });
+
+      if (!uploadResponse.ok) {
+        throw new Error("Failed to upload file to S3");
+      }
+
+      // Step 3: Save metadata to your database
+      const metadataResponse = await fetch("/api/recordedGames", {
```

Comment:
it's a very small optimisation that probably shouldn't be attempted until everything else is working, but since step 3 doesn't rely on the result of step 2 (unless I've missed something?), lines 68-93 can be rearranged to have both requests going at once rather than running one at a time:

```ts
      const uploadResponsePromise = fetch(signedUrl, {
        method: "PUT",
        body: recFile,
        headers: {
          "Content-Type": fileType,
        },
        mode: "cors",
      });


      // Step 3: Save metadata to your database
      const metadataResponsePromise = fetch("/api/recordedGames", { 
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          gameTitle: fileName,
          s3Key: key,
        }),
      });

      const uploadResponse = await uploadResponsePromise;

      if (!uploadResponse.ok) {
```

Reply:
Thanks for your review!

Yes we could run in parallel!However I rather not have the metadata be created if the file upload doesn't succeed. If that makes sense. 

I'm open to the idea though

Reply:
I would prefer them to stay sequential. When I was testing this, I was having issues where I would get a 200 response from the S3 PUT request, but the file was not being uploaded. Since we are doing this client side now, we may want to put into some thought as to what should happen if the rec upload or parse/mongo save fails. May not really even been necessary at the moment...but I did design things so that a single game guid could only be uploaded once (but I think I'm enforcing it from checking the mongo collection of parsed recs).